_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,185 @@
+This gives a very simple way of enforcing deep immutability.
+Removing a permission (and so constructing a deeply or shallowly immutable reference) is a simple register-register operation, less expensive in hardware than addition.
+
+The last property from the list, the ability to hand out tamper-proof opaque references, is implemented with CHERI's *sealing* mechanism.
+CHERI sealing associates an object type (a numerical value) with a pointer and, at the same time, makes the pointer unusable.
+Sealing requires a capability that authorises sealing with the specific type.
+Unsealing similarly requires a capability that authorises unsealing with a specific type.
+You can use this to enforce type safety for opaque types in the presence of unsafe code.
+
+I used this extensively in the [CHERI-JNI](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf) work back in 2017.
+When a Java program passed an object pointer to C, it was passed as a sealed capability.
+C code could do nothing with this other than pass it back to the Java VM via Java Native Interface (JNI) functions.
+The JNI exposes functions for getting or setting fields and calling methods.
+Each of these could unseal the object and know that it was a valid Java object.
+From there, there was no need for CHERI to be involved with type safety because every Java object carries a pointer to its class and so type checks were possible to implement purely in software.
+This highlights one of the core goals of CHERI: It *enables* languages to be safe, it doesn't *mandate* how they enforce that safety.
+The Java VM can efficiently enforce type safety (and therefore memory safety) within its own world without CHERI (though CHERI can improve performance in a few places), but CHERI enables it to retain these properties even when calling C.
+
+The remaining three properties are somewhat more specific to CHERIoT.
+Preventing an object from being accessed after it has been freed is possible on other CHERI systems but is implemented in different ways.
+[Cornucopia Reloaded](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/202404asplos-cornucopia-reloaded.pdf) explains an efficient way of implementing it on very large CHERI systems.
+CHERIoT takes an approach tailored for tiny embedded systems with no MMU.
+Each granule (8 bytes, by default) of memory that can be used for heap (which may not be all memory) has an associated bit in a bitmap.
+When an object is freed, the allocator sets the associated bits marking it as free.
+From this point on, you can never load a pointer to that object.
+Eventually, the revoker deletes pointers to that object from memory.
+After that has happened, the allocator can paint the bits in the 'freed' bitmap again and allow the memory to be reused.
+The monotonicity properties of CHERI mean that a pointer that points to an object can be used only to derive objects with equal or smaller bounds.
```

Comment:
That's a little out of left field, but I agree it's important to say.  Connect it back to revocation with something like...

And therefore, any capability derived from a pointer to a freed heap object (say, to an interior structure) will also be revoked.